ddtrace/contrib/sanic/patch.py

```diff
@@ -90,41 +99,59 @@ def _get_path(request):
 async def patch_run_request_middleware(wrapped, instance, args, kwargs):
     # Set span resource from the framework request
     request = args[0]
-    pin = Pin._find(request.ctx)
-    if pin is not None and pin.enabled():
-        span = pin.tracer.current_span()
-        if span is not None:
-            span.resource = "{} {}".format(request.method, _get_path(request))
+    span = _get_current_span(request)
+    if span is not None:
+        span.resource = "{} {}".format(request.method, _get_path(request))
     return await wrapped(*args, **kwargs)
 
 
 def patch():
     """Patch the instrumented methods."""
-    global SANIC_PRE_21
+    global SANIC_VERSION
 
     if getattr(sanic, "__datadog_patch", False):
         return
     setattr(sanic, "__datadog_patch", True)
 
-    SANIC_PRE_21 = sanic.__version__[:2] < "21"
+    SANIC_VERSION = tuple(map(int, sanic.__version__.split(".")))
 
-    _w("sanic", "Sanic.handle_request", patch_handle_request)
-    if not SANIC_PRE_21:
-        _w("sanic", "Sanic._run_request_middleware", patch_run_request_middleware)
+    if SANIC_VERSION >= (21, 9, 0):
+        _w("sanic", "Sanic.__init__", patch_sanic_init)
         _w(sanic.request, "Request.respond", patch_request_respond)
+    else:
+        _w("sanic", "Sanic.handle_request", patch_handle_request)
+        if SANIC_VERSION >= (21, 0, 0):
+            _w("sanic", "Sanic._run_request_middleware", patch_run_request_middleware)
+            _w(sanic.request, "Request.respond", patch_request_respond)
 
 
 def unpatch():
     """Unpatch the instrumented methods."""
-    _u(sanic.Sanic, "handle_request")
-    if not SANIC_PRE_21:
-        _u(sanic.Sanic, "_run_request_middleware")
-        _u(sanic.request.Request, "respond")
     if not getattr(sanic, "__datadog_patch", False):
         return
+
+    if SANIC_VERSION >= (21, 9, 0):
+        _u(sanic.Sanic, "__init__")
+        _u(sanic.request.Request, "respond")
+    else:
+        _u(sanic.Sanic, "handle_request")
+        if SANIC_VERSION >= (21, 0, 0):
+            _u(sanic.Sanic, "_run_request_middleware")
+            _u(sanic.request.Request, "respond")
+
     setattr(sanic, "__datadog_patch", False)
 
 
+def patch_sanic_init(wrapped, instance, args, kwargs):
+    """Wrapper for creating sanic apps to automatically add our signal handlers"""
+    wrapped(*args, **kwargs)
```

Comment:
I am not sure it makes sense here since if wrapped raises an exception our code won't get called/we won't have a valid `instance` to call `add_signal` on.

And if our code raises an exception... probably better to raise here on app init?